### PR TITLE
feat: Add toggle for editor line length per user

### LINKED ISSUE #4024: Configurable line-length for Editor
Hello there, 

I have had a look through previous issues regarding the same broad topic, but as those are all currently closed, I'd like to open one here again. My Feature request is about the current Limit on the amount of characters/pixels in one Line in the Editor.

Related Issues I found: #24  #125

I can understand the Aspect of this setting currently being at about 80–120 characters (or is it now based on pixels?) because experts say that it should be, but not everyone wants to publish his/her work and reasonably might prefer to use more or less than (in my case) 1/3 of the screen width on a 15" Full-HD landscape display.

**Describe the solution you'd like**
~I'd like to be able to set the line-length/padding either from the Admin Settings or by the User in the Personal Settings.~
Alternatively, a Slider on the Bottom of the Editor View that defaults to the currently fixed amount of characters and could be dynamically adjusted would be great.

**Describe alternatives you've considered**
Setting my browser to 130% Zoom, which only solves the Issue of "wasted" horizontal space, but not the Issue of not being able to fit more Information in one line (for example in Tables with more than 4 columns).

**Additional context**
Default View (:root --text-editor-max-with: 670px) at 100% Zoom
Bildschirmfoto vom 2023-04-04 23-16-00-obfuscated

Default View (:root --text-editor-max-with: 670px) at 133% Zoom
Bildschirmfoto vom 2023-04-04 23-29-44-obfuscated

Modified View (:root --text-editor-max-with: 1200px) at 100% Zoom
Bildschirmfoto vom 2023-04-04 23-27-02-obfuscated

Thank you very much!

**Post Edits:**
- Added physical Display Size
- Changed wording from amount of chars to padding
- crossed out "preferred" solution

### PR DESCRIPTION
**Pull request by azri-cs**
### 📝 Summary

Fixes #4024

After some direct alignment with designers, we basically want to offer a per user setting to allow users changing between a full line length and a compact line length display mode.

#### 🖼️ Screenshots

| 🏚️ Before | 🏡 After |
|--|--|
| default | full-width |

B | A

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [x] [Tests](https://github.com/nextcloud/text#-testing-the-app) (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation ([README](https://github.com/nextcloud/text/blob/main/README.md) or [documentation](https://github.com/nextcloud/documentation/blob/master/admin_manual/configuration_server/text_configuration.rst#L2)) has been updated or is not required

### Follow up

- [ ] Hide the toggle for line length in collectives as it has its own option
- [ ] Discuss if moving the toggle to the action menu makes sense

### CODE REVIEW

**Comment by max-nextcloud** [comment]
Hi @azri-cs :wave: 

Looks like this is your first PR to text. :tada: 

:heart: Thanks a lot for contributing! I hope we can walk you through finalizing this.
I assume you are addressing #4024 - so I'll add that to the PR description.

First of all, let's check in with @nextcloud/designers to hear if the ui matches their expectations.

In the meantime we can try and address the failing tests:

### DCO

The DCO check is complaining about your signoff message in the commits like this:
grafik

If you use git on the command line you can add a signoff message to the last commit with
`git commit --amend --signoff`.


### Lint PHP-cs

At first sight this looks like it's mostly about indentation. Can you see the error message the workflow printed? It has a detailed diff of what needs to be changed.

### Psalm

This also looks related:

grafik


I hope these are helpful pointers to address these. If you get stuck feel free to ask. I will also take a first look at the code to see if I find something to comment upon.

File: src/components/Editor.vue
Comment (max-nextcloud) [inline src/components/Editor.vue:1]: The whitespace of this file somehow changed so that your commit seems to change every line inside the file.
This would be confusing when looking at the history for a given line.

Please try to leave the whitespace untouched. In the server repo there is a .editorconfig file that will tell most editors how we like our indentation.
This may not have any effect for two reasons:
1. You may need an editorconfig extensions / configuration setting to make sure your editor reads `.editorconfig` files.
2. The `.editorconfig` file needs ot be in a parent directory of the text repository. Maybe your text repo is not located inside your server repo. I also have that situation. `text` lives in `code/nextcloud/text` on my machine. So I copied the `.editorconfig` to `code/nextcloud` where it will also affect text.

**Review by max-nextcloud — COMMENTED** [review]
Code wise this looks good.

I noticed that you added a string config setting - but I think a boolean would do. If I read #4024 correctly it's either full width or default line length.

File: src/components/Editor.vue
Comment (azri-cs) [reply src/components/Editor.vue:1]: Hi Max,

Thank you for pointing that out. I have pushed the fixes for sign-off issue, whitespaces and backend test failure.

**Comment by azri-cs** [comment]
> Code wise this looks good.
> 
> I noticed that you added a string config setting - but I think a boolean would do. If I read #4024 correctly it's either full width or default line length.

Upon reading back the requirements, yes I agree boolean is better since it's a checkbox and there are only 1 option and 1 default. My latest push have implemented this.

**Comment by max-nextcloud** [comment]
Dear @azri-cs Thanks for changing the approach and using a boolean. That looks good to me.

The linters are still not quite happy. They mostly disagree about the use of whitespace here and there.
For the javascript side of things you can run the linter locally with:
`npm run lint`
It looks like all the errors it reports can be fixed automatically. You can do this by running
`npx eslint --fix --ext .js,.vue src`

For php i'm less familiar but my understanding is that you will need to run `composer install` first to get all the composer dependencies and then you can run what the CI job also runs:
`composer run cs:check` for checking and
`composer run cs:fix` for fixing.

**Comment by azri-cs** [comment]
> Dear @azri-cs Thanks for changing the approach and using a boolean. That looks good to me.
> 
> The linters are still not quite happy. They mostly disagree about the use of whitespace here and there. For the javascript side of things you can run the linter locally with: `npm run lint` It looks like all the errors it reports can be fixed automatically. You can do this by running `npx eslint --fix --ext .js,.vue src`
> 
> For php i'm less familiar but my understanding is that you will need to run `composer install` first to get all the composer dependencies and then you can run what the CI job also runs: `composer run cs:check` for checking and `composer run cs:fix` for fixing.

Hi @max-nextcloud

Sorry for the inconvenience caused. I've followed your advice to use `.editorconfig` from the server repo, and then I run `ctrl+alt+l` to format each files I previously committed. Not sure why the whitespaces issue still exists despite of this. I've fix the whitespaces issue on my latest push.

**Comment by max-nextcloud** [comment]
Hey @azri-cs 
Thanks for following up on this so swiftly. I will ping the design team again for review. Code wise i think this is ready to be merged.
There's a signature missing on one of the commits - but I think it would make sense to squash all of the commits into one anyway so the back and forth with the whitespace changes does not show up in history.

Could you squash the commits?
Feedback from the design team might not arrive before Monday next week as it's the weekend for most of them now.

**Comment by azri-cs** [comment]
> Hey @azri-cs Thanks for following up on this so swiftly. I will ping the design team again for review. Code wise i think this is ready to be merged. There's a signature missing on one of the commits - but I think it would make sense to squash all of the commits into one anyway so the back and forth with the whitespace changes does not show up in history.
> 
> Could you squash the commits? Feedback from the design team might not arrive before Monday next week as it's the weekend for most of them now.

Hi @max-nextcloud 

No worries. I've squashed my commits into one commit. Thanks a lot, I've learnt a lot.

**Review by jancborchardt — COMMENTED** [review]
Very nice contribution @azri-cs! :) I only have 1 feedback: I would expect a setting like this to be in the collapsed part of the toolbar along with these other more info/settings like items, as the "people" menu is more focused on collaboration and sharing.
(Also, this setting is disabled by default, right?)
image

**Comment by jancborchardt** [comment]
@juliusknorr @azri-cs sure, doing it as a follow-up is fine for me too. :)

My remark does make sense though right, or would you still prefer your originally suggested placement?

**Review by juliusknorr — APPROVED** [review]
The code looks good and works like a charm. Let's get this in as a first iteration and I'll file follow up issues for polishing.

**Comment by juliusknorr** [comment]
> My remark does make sense though right, or would you still prefer your originally suggested placement?

I'm undecided on that part as currently all the menu items are just buttons and having a global toggle there seems out of place a bit, but I also see that the "collaborator" menu is not ideal as well.

**Comment by juliusknorr** [comment]
@azri-cs Thanks for the fixes as well. Seems we have a conflicting file meanwhile. Could you rebase your changes against the latest main branch and resolve the conflict?

**Comment by juliusknorr** [comment]
Collected follow up topics in the pr description to file in separate issues after merging the first step

**Comment by azri-cs** [comment]
> @azri-cs Thanks for the fixes as well. Seems we have a conflicting file meanwhile. Could you rebase your changes against the latest main branch and resolve the conflict?

Sorry @juliusknorr , I'm facing few issues:
1. Upon resolving the conflict via Github.com it created a merge commit from main with my branch. It was because the conflict doesn't appear on my local even when I run `git rebase main` hence why I used Github.com instead.
2. And then afterwards, I just realized I missed one line of code and pushed a new commit as a fix.

Are these okay? If not, what should I do?

**Comment by juliusknorr** [comment]
Would be great if you could rebase and squash it into one commit locally. Unfortunately GitHub doesn't allow that with the conflict reduction web ui

**Comment by max-nextcloud** [comment]
@azri-cs 
> Is this correct? How can I resolve the new conflicting files as I did not receive the conflict files locally to solve even when running `git rebase origin/main`?

Looks like something with the earlier squash went wrong. Now the branch includes commits that have already been on main branch etc.

The last good commit seems to have been https://github.com/nextcloud/text/commit/e67590af683d7d85c9c7c49023aa4cea8abd4e0e (before https://github.com/nextcloud/text/pull/6569#issuecomment-2452693172)

So I suggest you start again from there. Could you try if this works for you:
```bash
git reset --hard e67590af683d7d85c9c7c49023aa4cea8abd4e0e
```

It does not work for me - but that might be because I do not have that commit in my own git repo as it only was on your repo.

**Comment by juliusknorr** [comment]
One additional hint your origin remote is probably your fork with is behind the upstream repository. Mabye also try updating your fork with the latest changes on the main branch at https://github.com/azri-cs/text

**Comment by juliusknorr** [comment]
Did another rebase myself as somehow CI was not running

**Comment by max-nextcloud** [comment]
> Did another rebase myself as somehow CI was not running

As this is on a fork the CI run needs to be approved. Did that.

**Comment by juliusknorr** [comment]
Interesting, I didn't see that with the new merge UI that I've enabled as a preview

**Pull request merged**
